### PR TITLE
Bug 1999119: bump to golang-1.16 

### PR DESCRIPTION
**Pull request by eslutsky**
We need the support of the new features in the component.

Signed-off-by: Evgeny Slutsky <eslutsky@redhat.com>

### CODE REVIEW

**Comment by eslutsky** [comment]
this change also requires bumping the golang build container image to 1.16

**Comment by patrickdillon** [comment]
If we are going to bump golang version in this PR it should be a separate commit.

**Comment by eslutsky** [comment]
> If we are going to bump golang version in this PR it should be a separate commit.

done

**Review by patrickdillon — COMMENTED** [review]
For 1.16 this should also update `hack/build.sh` and the dockerfiles. If this turns out to be too much work let me know and we can open a separate pr

File: go.sum [hunk 1]
```diff
@@ -490,7 +490,6 @@ github.com/elazarl/goproxy v0.0.0-20180725130230-947c36da3153/go.mod h1:/Zj4wYkg
 github.com/elazarl/goproxy v0.0.0-20190911111923-ecfe977594f1 h1:yY9rWGoXv1U5pl4gxqlULARMQD7x0QG85lqEXTWysik=
 github.com/elazarl/goproxy v0.0.0-20190911111923-ecfe977594f1/go.mod h1:Ro8st/ElPeALwNFlcTpWmkr6IoMFfkjXAvTHpevnDsM=
 github.com/elazarl/goproxy/ext v0.0.0-20190711103511-473e67f1d7d2/go.mod h1:gNh8nYJoAm43RfaxurUnxr+N1PwuFV3ZMl/efxlIlY8=
-github.com/elazarl/goproxy/ext v0.0.0-20190911111923-ecfe977594f1 h1:8B7WF1rIoM8H1smfpXFvOawSAzlRDMVzoGu9zE3+OCk=
```
[the remark this answers is not in the document]
Comment (eslutsky) [reply go.sum]: done

File: go.sum [hunk 2]
```diff
@@ -1345,7 +1344,6 @@ github.com/nats-io/nats.go v1.9.1/go.mod h1:ZjDU1L/7fJ09jvUSRVBR2e7+RnLiiIQyqyzE
 github.com/nats-io/nkeys v0.1.0/go.mod h1:xpnFELMwJABBLVhffcfd1MZx6VsNRFpEugbxziKVo7w=
 github.com/nats-io/nkeys v0.1.3/go.mod h1:xpnFELMwJABBLVhffcfd1MZx6VsNRFpEugbxziKVo7w=
 github.com/nats-io/nuid v1.0.1/go.mod h1:19wcPz3Ph3q0Jbyiqsd0kePYG7A95tJPxeL+1OSON2c=
-github.com/nbio/st v0.0.0-20140626010706-e9e8d9816f32 h1:W6apQkHrMkS0Muv8G/TipAy/FJl/rCYT0+EuS8+Z0z4=
```
[the remark this answers is not in the document]
Comment (eslutsky) [reply go.sum]: done

File: go.sum [hunk 3]
```diff
@@ -1936,7 +1934,6 @@ go.uber.org/multierr v1.1.0/go.mod h1:wR5kodmAFQ0UK8QlbwjlSNy0Z68gJhDJUG5sjR94q/
 go.uber.org/multierr v1.3.0/go.mod h1:VgVr7evmIr6uPjLBxg28wmKNXyqE9akIJ5XnfpiKl+4=
 go.uber.org/multierr v1.5.0 h1:KCa4XfM8CWFCpxXRGok+Q0SS/0XBhMDbHHGABQLvD2A=
 go.uber.org/multierr v1.5.0/go.mod h1:FeouvMocqHpRaaGuG9EjoKcStLC43Zu/fmqdUMPcKYU=
-go.uber.org/tools v0.0.0-20190618225709-2cfd321de3ee h1:0mgffUl7nfd+FpvXMVz4IDEaUSmT1ysygQC7qYo7sG4=
```
[the remark this answers is not in the document]
Comment (eslutsky) [reply go.sum]: done

**Comment by eslutsky** [comment]
> For 1.16 this should also update `hack/build.sh` and the dockerfiles. If this turns out to be too much work let me know and we can open a separate pr

it does get bigger  .. lets separate this PR from bumping the cluster-api-provider which depends on it.

**Comment by patrickdillon** [comment]
This still needs to update: docs/dev/dependencies.md  and hack/build.sh 

I am using https://github.com/openshift/installer/commit/46a5f49709d48a5241ee87962c24163f322d8267 as a reference

**Comment by patrickdillon** [comment]
Everything else looks good. I think the changes to `go.sum` must be an artifact of the upgrade. I tried it out locally.

**Comment by eslutsky** [comment]
> This still needs to update: docs/dev/dependencies.md and hack/build.sh
> 
> I am using [46a5f49](https://github.com/openshift/installer/commit/46a5f49709d48a5241ee87962c24163f322d8267) as a reference

done

**Comment by eslutsky** [comment]
> Everything else looks good. I think the changes to `go.sum` must be an artifact of the upgrade. I tried it out locally.

done

**Comment by LorbusChris** [comment]
One should be able to fix the failing `verify-vendor` test by running (with a Go 1.16 binary) and committing the output of `go mod tidy && go mod vendor`  on this branch.

**Comment by patrickdillon** [comment]
> > Everything else looks good. I think the changes to `go.sum` must be an artifact of the upgrade. I tried it out locally.
> 
> done

Apologies if my quoted comment was unclear. I was trying to say that `go.sum` should indeed be included.  If you re-include it in your pr, verify vendor should pass.

/approve

**Pull request merged**